include/deal.II/base/tensor_accessors.h

```diff
@@ -0,0 +1,586 @@
+   */
+  template <typename T>
+  struct AccessType
+  {
+    typedef typename T::value_type value_type;
+  };
+
+
+  template <typename T>
+  struct AccessType<const T>
+  {
+    typedef const typename T::value_type value_type;
+  };
+
+
+  /**
+   * This class provides a local typedef @p value_type that is equal to the
+   * typedef <code>value_type</code> after @tparam rank - 1 recursive
+   * applications. Further, constness is preserved via the AccessType type
+   * trait, i.e., if T is const, ReturnType<rank, T>::value_type will also
+   * be const.
+   *
+   * @author Matthias Maier, 2015
+   */
+  template <int rank, typename T>
+  struct ReturnType
+  {
+    typedef typename ReturnType<rank - 1, typename AccessType<T>::value_type>::value_type value_type;
+  };
+
+  template <typename T>
+  struct ReturnType<0, T>
+  {
+    typedef T value_type;
+  };
+
+
+  namespace internal
+  {
+    // OK. This is utterly brutal template magic.
+    // Best to start reading at ReorderedIndexView...
+
+    template <int rank, typename S> class StoreIndex;
+    template <int index, int rank, typename T> class ReorderedIndexView;
+
+    // A type trait to allow nested application of the index reordering
+    // classes. The problem is that when working with the actual tensorial
+    // types, we have to return subtensors by reference - but StoreIndex
+    // and ReorderedIndexView have to be returned by value (those are
+    // rvalues...)
+    template<typename T>
+    struct ReferenceType
+    {
+      typedef T &type;
+    };
+
+    template <int rank, typename S>
+    struct ReferenceType<StoreIndex<rank, S> >
+    {
+      typedef StoreIndex<rank, S> type;
+    };
+
+    template <int index, int rank, typename T>
+    struct ReferenceType<ReorderedIndexView<index, rank, T> >
+    {
+      typedef ReorderedIndexView<index, rank, T> type;
+    };
+
+
+    // A type trait that strips StoreIndex<0, S> down to its actual return
+    // value. This is needed to end the recursion in StoreIndex as well as,
+    // to return something meaningful in case of a nested application of
+    // the index reordering classes.
+    template<typename T>
+    struct StripStoreIndex
+    {
+      typedef T type;
+    };
+
+    template <typename S>
+    struct StripStoreIndex<StoreIndex<0, S> >
+    {
+      typedef typename StripStoreIndex<typename S::return_type>::type type;
+    };
+
+
+    // A class that stores an index recursively with every invocation of
+    // operator[](unsigned int). After all rank indices are stored, the
+    // specialization StoreIndex<0, typename S> is returned that will apply
+    // the stored indices to the underlying tensor object with the first
+    // index moved to the end, i.e., an invocation
+    // @code
+    //   StoreIndex<... Tensor ...>(... t) [i_0][i_1]...[i_n]
+    // @endcode
+    // results in the following object being returned:
+    // @code
+    //   t[i_1]...[i_n][i_0]
+    // @endcode
```

Comment:
This is bizarre :-) Where would you need this kind of behavior? Can you elaborate why one would want this rotation of indices?